compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -1295,9 +1293,26 @@ object Parsers {
         case _: (ForYield | ForDo) => in.token == FOR
         case _ => false
 
+      def matchesAndSetEnd(stat: T): Boolean = {
+        val didMatch = matches(stat)
+        if didMatch then
+          stat match
+            case stat: WithEndMarker =>
+              stat.withEndMarker()
+            case _ =>
+              ()
+
+          // VERY SNEAKY MUTATION HERE:
+          // `withSpan` can clone the tree, so we must replace the last statment
```

Comment:
Instead of withSpan, could you use span_= which doesn't mutate? https://github.com/lampepfl/dotty/blob/fa3657708183167e9d927f4118065aabcddabb79/compiler/src/dotty/tools/dotc/ast/Positioned.scala#L46

Reply:
thank you for the suggestion